archive/2025-11-07-you-should-write-an-agent-the-fly-blog.md

```diff
@@ -0,0 +1,299 @@
+
+So, just like you, I rolled my eyes when “Prompt Engineering” turned into “Context Engineering”. Then I wrote an agent. Turns out: context engineering is a straightforwardly legible programming problem.
+
+You’re allotted a fixed number of tokens in any context window. Each input you feed in, each output you save, each tool you describe, and each tool output eats tokens (that is: takes up space in the array of strings you keep to pretend you’re having a conversation with a stateless black box). Past a threshold, the whole system begins getting nondeterministically stupider. Fun!
+
+No, really. Fun! You have so many options. Take “sub-agents”. People make a huge deal out of Claude Code’s sub-agents, but you can see now how trivial they are to implement: just a new context array, another `call` to the model. Give each `call` different tools. Make sub-agents talk to each other, summarize each other, collate and aggregate. Build tree structures out of them. Feed them back through the LLM to summarize them as a form of on-the-fly compression, whatever you like.
+
+Your wackiest idea will probably (1) work and (2) take 30 minutes to code.
+
+Haters, I love and have not forgotten about you. You can think all of this is ridiculous because LLMs are just stochastic parrots that hallucinate and plagiarize. But what you can’t do is make fun of “Context Engineering”. If Context Engineering was an [Advent of Code problem](https://adventofcode.com/ ""), it’d occur mid-December. It’s programming.
+
+## [Anchor](https://fly.io/blog/everyone-write-an-agent/\#nobody-knows-anything-yet-and-it-rules) Nobody Knows Anything Yet And It Rules
+
+Maybe neither will! Skeptics could be right. ( [Seems unlikely though](https://www.darpa.mil/research/programs/ai-cyber "").)
+
+[Startups have raised tens of millions](https://xbow.com/ "") building agents to look for vulnerabilities in software. I have friends doing the same thing alone in their basements. Either group could win this race.
+
+I am not a fan of the OWASP Top 10.
+
+I’m stuck on vulnerability scanners because I’m a security nerd. But also because it crystallizes interesting agent design decisions. For instance: you can write a loop feeding each file in a repository to an LLM agent. Or, as we saw with the ping example, you can let the LLM agent figure out what files to look at. You can write an agent that checks a file for everything in, say, the OWASP Top 10. Or you can have specific agent loops for DOM integrity, SQL injection, and authorization checking. You can seed your agent loop with raw source content. Or you can build an agent loop that builds an index of functions across the tree.
+
+You don’t know what works best until you try to write the agent.
```

Comment:
This is an important point. Tests with agents aren't coding tests. They are data science tests--evals.